### ISSUE TITLE
What's the behavior with already serialized data in production ?

### ISSUE THREAD

**Issue by bgaillard** [igbinary/igbinary#118]
Hi, we would like to enable igbinary on a production environnent having the following configuration.

```
igbinary support => no
memcached support => enabled
memcached.serializer => php => php
apc.serializer => php => php
session.serialize_handler => php => php
``` 

We wish to change our configuration to.

```
igbinary support => yes
memcached support => enabled
memcached.serializer => igbinary => igbinary
apc.serializer => igbinary => igbinary
session.serialize_handler => igbinary => igbinary
``` 

My question is related to existing serialized data in our production system. 

Currently we have lots of data in our Memcached servers which have been serialized with the `php` serializer (those data are mostly session data and other application specific caching stuffs).

What will be the behavior if we change the serializer to `igbinary` now ? 

* Do we have to cleanup all our Memcached caches and then restart with fresh caches containing only `igbinary` serialized data (in this case its not cool for the users which have session cookies) ?
* Or, is `igbinary` able to detect if data have been serialized with the `php` serializer before ? In this case is `igbinary` able to unserialize the data with the `php` serializer and "re-serialize" those data with `igbinary` ? 

Thanks for your help.

Baptiste

**Comment by TysonAndre** [comment]
Memcached stores the serializer it used at the time of serialization. (Not sure what version it started doing that in).

By memcached, I assume you mean https://github.com/php-memcached-dev/php-memcached ?

There is an enum for flags associated with the values
added on serialization (uses memcache flags on keys): https://github.com/php-memcached-dev/php-memcached/blob/master/php_memcached.c#L918
and checked on unserialization: https://github.com/php-memcached-dev/php-memcached/blob/master/php_memcached.c#L3455


If you're using settings like compression, there have recently been bugs where the flags were stored incorrectly if you used a custom serializer and compression at the same time. I didn't check which version they were introduced/fixed in.

You can set up a test server to be absolutely sure what the behavior will be.

**Comment by TysonAndre** [comment]
And for APCu:

- Restart the web server(e.g. apache) after changing the serializer setting
- deploy config changes for APCu separately from memcached APCu change

**Comment by bgaillard** [comment]
Hi @TysonAndre and thanks for your reactivity on this case. 

> By memcached, I assume you mean https://github.com/php-memcached-dev/php-memcached ?

Yes we are using a recent version of `php-memcached` and an Amazon AWS Elasticache cluster. 

Ok, so as you explained everything should work in our case, thanks for the explanations about the flags used at serialization / unserialization.

> You can set up a test server to be absolutely sure what the behavior will be.

We've finally decided to update today. We've cleared the most sensible memory caches to persist their values in our database, we took the risk for other less sensible data.

Finally for APC we restarted all our Web Servers to restart from fresh caches.

Everything went ok except that we encountered the following error.

```
session_start(): Cannot find serialization handler 'igbinary' - session startup failed
```

For now we set back the `session.serialize_handler` to `php`, don't know why the `igbinary` serializer cannot be found only for sessions (we have no errors for Memcached and APC).
